### PR TITLE
Function return types

### PR DESCRIPTION
**Pull request by benjie**
Hey again; this time I actually have a proof of concept!

I thought it would be nice to extend massive so that when you call an SQL function that returns, say, an integer; that you get an integer result. e.g. if you had this function:

```sql
create function myschema."getRandomNumber"()
returns integer as
$$
select 4; -- chosen by fair dice roll.
          -- guaranteed to be random.
$$ language sql;
```

Currently calling that in massive will give `[{getRandomNumber: 4}]`:

```js
db.myschema.getRandomNumber((err, res) => {
  assert.deepEqual(res, [{getRandomNumber: 4}]);
});
```

This PR changes that such that you instead get simply `4`:

```js
db.myschema.getRandomNumber((err, res) => {
  assert.equal(res, 4);
});
```

I've done it in as minimally invasively a way as I can achieve; but that means that anyone relying on the old behaviour (as I had started to do in my own project) it will break for. I think we can get around this easily by just allowing people to opt-in to this feature via a configuration parameter, but I've not gone as far as to add that without receiving feedback from you first.

What are your thoughts?

---

*🎩 HT XKCD 221*

### CODE REVIEW

File: lib/executable.js [hunk 1]
```diff
@@ -10,6 +10,12 @@ var Executable = function(args) {
 
   this.sql = args.sql;
   this.filePath = args.filePath;
+  this.singleRow = args.singleRow;
+  this.singleValue = args.singleValue;
+
+  if (this.singleValue && !this.singleRow) {
+    throw new Error("We do not yet support singleValue unless singleRow is set.");
+  }
```
Comment (benjie) [inline lib/executable.js]: Whoops, these should be asserts really; I was just being lazy.

**Comment by dmfay** [comment]
Very cool! Couple of comments:

* `invoke` should bubble the result errors instead of throwing. The throw for trying to stream is one thing since it speaks to a problem with how the function is being invoked in the first place, but once it's actually executed we defer to the client's error handling.
* As far as I know there's nothing in the way of supporting domain, enum, or range types (`d`, `e`, & `r`) in addition to base types.

**Comment by benjie** [comment]
Domains, enums and arrays support added (where possible within the limitations of the `pg` module); have disabled interpreting by default so as to not break backwards compatibility (a suggestion on what to name this would be welcome!); and have turned the throws into bubbles.

**Comment by benjie** [comment]
@dmfay I've taken heed of your feedback, fixed those issues and now added support for `setof text`-style return values (as well as `setof row` which was already supported); and have added support for streaming them too.

**Comment by dmfay** [comment]
Code looks good, can you add some documentation to functions.md under 'Function Invocation'?

**Comment by benjie** [comment]
Done; I also added a note about it in `connecting.md` and changed the configuration variable from the unwieldy `interpretFunctionReturnTypes` to simply `enhancedFunctions` - I hope that's okay?

**Comment by dmfay** [comment]
that works, thanks!

**Pull request merged**